File: frontend/types/message.ts
// 群组消息
export interface GroupMessage {
  message_id: string
  group_id: string
  sender_id: number
  sender_name: string
  content: string
  message_type: string
  sent_at: string
  latitude: number
  longitude: number
}

// 私聊消息
export interface DirectMessage {
  message_id: string
  conversation_id: string
  sender_id: number
  sender_name: string
  recipient_id: number
  recipient_name: string
  content: string
  message_type: string
  sent_at: string
  read_at: string | null
  latitude: number
  longitude: number
}

// 会话信息
export interface Conversation {
  conversation_id: string
  peer_id: number
  peer_name: string
  last_message: string
  last_message_time: string
  unread_count: number
}

// 发送群组消息请求
export interface SendGroupMessageRequest {
  group_id: string
  content: string
  message_type?: string
  latitude: number
  longitude: number
}

// 发送群组消息响应
export interface SendGroupMessageResponse {
  message_id: string
  group_id: string
  message_type: string
  sent_at: string
}

// 获取群组消息历史请求
export interface GroupMessageHistoryRequest {
  group_id: string
  cursor?: number
  limit?: number
}

// 发送私聊消息请求
export interface SendDirectMessageRequest {
  recipient_id: number
  content: string
  message_type?: string
  latitude: number
  longitude: number
}

// 发送私聊消息响应
export interface SendDirectMessageResponse {
  message_id: string
  conversation_id: string
  recipient_id: number
  message_type: string
  sent_at: string
}

// 获取私聊消息历史请求
export interface DirectMessageHistoryRequest {
  conversation_id?: string
  user_id?: number
  limit?: number
  cursor?: number
}

// 获取会话列表请求
export interface ConversationsRequest {
  limit?: number
  cursor?: number
}

// 标记消息已读请求
export interface MarkReadRequest {
  conversation_id: string
  message_id?: string
}

// 标记消息已读响应
export interface MarkReadResponse {
  success: boolean
  marked_count?: number
}

// 删除私聊消息请求
export interface DeleteDirectMessageRequest {
  message_id: string
}

// 删除消息响应
export interface DeleteMessageResponse {
  success: boolean
}

// 分页响应结构
export interface PaginatedResponse<T = any> {
  items: T[]
  has_more: boolean
  next_cursor?: number | null
} 

File: backend/src/models/api/message.rs
use crate::utils::response::{PaginatedResponse, Pagination};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

// ============ 群组消息相关请求/响应模型 ============

#[derive(Debug, Serialize, Deserialize)]
pub struct SendGroupMessageRequest {
    pub group_id: String,
    pub content: String,
    pub message_type: String, // 默认为 "text"，可以是 "image" 等
    pub latitude: f64,
    pub longitude: f64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SendGroupMessageResponse {
    pub group_id: String,
    pub message_id: String,
    pub message_type: String,
    pub sent_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GroupMessageHistoryRequest {
    pub group_id: String,
    #[serde(flatten)]
    pub pagination: Pagination,
}

pub type GroupMessageHistoryResponse = PaginatedResponse<GroupMessageHistoryItem>;

#[derive(Debug, Serialize, Deserialize)]
pub struct DeleteGroupMessageRequest {
    pub message_id: String,
}

// ============ 私聊消息相关请求/响应模型 ============

#[derive(Debug, Serialize, Deserialize)]
pub struct SendDirectMessageRequest {
    pub recipient_id: String,
    pub content: String,
    pub message_type: String, // 默认为 "text"，可以是 "image" 等
    pub latitude: f64,
    pub longitude: f64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SendDirectMessageResponse {
    pub message_id: String,
    pub conversation_id: String,
    pub recipient_id: String,
    pub message_type: String,
    pub sent_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DirectMessageHistoryRequest {
    pub conversation_id: Option<String>,
    pub user_id: Option<i64>,
    #[serde(flatten)]
    pub pagination: Pagination,
}

pub type DirectMessageHistoryResponse = PaginatedResponse<DirectMessageHistoryItem>;

#[derive(Debug, Serialize, Deserialize)]
pub struct DeleteDirectMessageRequest {
    pub message_id: String,
}

/// 标记消息已读请求
/// - conversation_id: 必需，会话ID
/// - message_id: 可选，指定消息ID。如果提供，则只标记该消息及之前的所有消息为已读
#[derive(Debug, Serialize, Deserialize)]
pub struct MarkReadRequest {
    /// 会话ID，必需
    pub conversation_id: String,
    /// 消息ID，可选。如果提供，则只标记该消息及之前的所有消息为已读
    pub message_id: Option<String>,
}

/// 标记消息已读响应
#[derive(Debug, Serialize, Deserialize)]
pub struct MarkReadResponse {
    /// 操作是否成功
    pub success: bool,
    /// 标记为已读的消息数量
    pub marked_count: Option<i64>,
}

// ============ 通用响应模型 ============

#[derive(Debug, Serialize, Deserialize)]
pub struct DirectMessageHistoryItem {
    pub message_id: String,
    pub conversation_id: String,
    pub sender_id: String,
    pub sender_name: String,
    pub recipient_id: String,
    pub content: String,
    pub message_type: String,
    pub sent_at: DateTime<Utc>,
    pub read_at: Option<DateTime<Utc>>,
    pub latitude: f64,
    pub longitude: f64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct GroupMessageHistoryItem {
    pub message_id: String,
    pub group_id: String,
    pub sender_id: String,
    pub sender_name: String,
    pub content: String,
    pub message_type: String,
    pub sent_at: DateTime<Utc>,
    pub latitude: f64,
    pub longitude: f64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DeleteMessageResponse {
    pub success: bool,
}

// ============ 会话列表相关模型 ============

#[derive(Debug, Serialize, Deserialize)]
pub struct ConversationsRequest {
    #[serde(flatten)]
    pub pagination: Pagination,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ConversationItem {
    pub conversation_id: String,
    pub peer_id: i64,
    pub peer_name: String,
    pub last_message_preview: String,
    pub last_message_time: DateTime<Utc>,
    pub unread_count: i64,
}

pub type ConversationsResponse = PaginatedResponse<ConversationItem>;
